File: kotlin/src/test/kotlin/org/libphonenumber/ae/kotlin/ConformanceTest.kt
package org.libphonenumber.ae.kotlin

import org.libphonenumber.ae.CountryCodeSource
import org.libphonenumber.ae.Format
import org.libphonenumber.ae.Leniency
import org.libphonenumber.ae.MatchType
import org.libphonenumber.ae.NumberType
import org.libphonenumber.ae.ValidationResult

/**
 * The 34-check binding conformance suite (docs/conformance.md, v2), in Kotlin.
 *
 * Proves the **Kotlin layer** reaches the same engine behaviour the Java and
 * Python suites see. Since that layer sits on the Java binding rather than on
 * its own FFI, what this really pins down is that the Kotlin sugar — the
 * top-level functions, the [Format]-keyed `format`, the `regions` property, the
 * re-exported value types — marshals every value shape correctly.
 *
 * It is NOT a phone-number test suite — the behavioural cases live in the
 * engine's own tests and run once, in Aether.
 *
 * A plain main method, not JUnit, for the same reason the Java suite is: the run
 * then needs nothing but a JDK and a Kotlin compiler, so it works offline and
 * cannot fail resolving a test-framework artifact.
 */
object ConformanceTest {

    private var passed = 0
    private val failures = mutableListOf<String>()

    @JvmStatic
    fun main(args: Array<String>) {
        // ---- the 34 required checks ----
        check("01 countryCode US") { assertEquals("1", countryCode("US")) }
        check("02 countryCode GB") { assertEquals("44", countryCode("GB")) }
        check("03 unknown region") { assertEquals("", countryCode("ZZ")) }
        check("04 exampleNumber US") { assertEquals("2015550123", exampleNumber("US")) }
        check("05 possibleLengths US") { assertEquals("10", possibleLengths("US")) }
        check("06 regionCodeForCountryCode 44") { assertEquals("GB", regionCodeForCountryCode("44")) }
        check("07 isNanpaCountry US") { assertTrue("US NANPA", isNanpaCountry("US")) }
        check("08 region enumeration") {
            val regs = regions
            assertTrue("region count >= 200, got ${regs.size}", regs.size >= 200)
            assertTrue("region[0] is 2 letters, got '${regs[0]}'", regs[0].length == 2)
        }
        check("09 regionsForCountryCode 1") {
            assertEquals("US", regionsForCountryCode("1")[0])
        }

        val parsed = parse("+1 201 555 0123 ext 42", "US")
        check("10 parse nationalNumber") { assertEquals("2015550123", parsed.nationalNumber()) }
        check("11 parse extension") { assertEquals("42", parsed.extension()) }
        check("12 parse countryCode") { assertEquals("1", parsed.countryCode()) }
        check("13 parse source FROM_NUMBER_WITH_PLUS") {
            assertEquals(CountryCodeSource.FROM_NUMBER_WITH_PLUS, parsed.source())
        }
        check("14 parse regionCode") { assertEquals("US", parsed.regionCode()) }

        check("15 parse GB trunk 0 stripped") {
            assertEquals("1212345678", parse("01212345678", "GB").nationalNumber())
        }
        check("16 isPossibleNumber US") { assertTrue("possible", isPossibleNumber("US", "2015550123")) }
        check("17 isPossibleWithReason TOO_SHORT") {
            assertEquals(ValidationResult.TOO_SHORT, isPossibleNumberWithReason("US", "201555"))
        }
        check("18 isValidNumber US") { assertTrue("valid", isValidNumber("US", "2015550123")) }
        check("19 isValidNumber wrong shape") {
            assertTrue("invalid", !isValidNumber("US", "1015550123"))
        }
        check("20 isValidNumber with +cc") { assertTrue("valid", isValidNumber("US", "+12015550123")) }
        check("21 numberType FIXED_LINE") {
            assertEquals(NumberType.FIXED_LINE, numberType("US", "2015550123"))
        }
        check("22 format NATIONAL") {
            assertEquals("(201) 555-0123", format("US", "2015550123", Format.NATIONAL))
        }
        check("23 format E164") {
            assertEquals("+12015550123", format("US", "2015550123", Format.E164))
        }
        check("24 format INTERNATIONAL") {
            assertEquals("+1 201-555-0123", format("US", "2015550123", Format.INTERNATIONAL))
        }
        check("25 format RFC3966") {
            assertEquals("tel:+1-201-555-0123", format("US", "2015550123", Format.RFC3966))
        }
        check("26 isNumberMatch EXACT") {
            assertEquals(MatchType.EXACT_MATCH, isNumberMatch("+12015550123", "+1 201 555 0123"))
        }
        check("27 isNumberMatch NO_MATCH") {
            assertEquals(MatchType.NO_MATCH, isNumberMatch("+12015550123", "+12025550123"))
        }
        check("28 normalizeDigitsOnly") {
            assertEquals("12015550123", normalizeDigitsOnly("+1 (201) 555.0123"))
        }
        check("29 convertAlphaCharacters") {
            assertEquals("1-800-3569377", convertAlphaCharacters("1-800-FLOWERS"))
        }
        check("30 truncateTooLong") {
            assertEquals("2015550123", truncateTooLong("US", "20155501239999"))
        }
        check("31 AsYouType") {
            val f = AsYouType("US")
            var out = ""
            for (c in "2015550123") out = f.inputDigit(c)
            assertEquals("(201) 555-0123", out)
        }
        check("32 findNumbers count") {
            val ms = findNumbers("call 201-555-0123 or +1 202 555 0199", "US", Leniency.VALID)
            assertEquals(2, ms.size)
        }
        check("33 findNumbers raw") {
            val ms = findNumbers("call 201-555-0123 now", "US", Leniency.VALID)
            assertEquals("201-555-0123", ms[0].raw())
        }
        check("34 abiVersion") { assertEquals(2, abiVersion()) }

        // ---- extras specific to the Kotlin layer ----
        check("format default style is NATIONAL") {
            assertEquals("(201) 555-0123", format("US", "2015550123"))
        }
        check("format helpers agree") {
            assertEquals("+12015550123", formatE164("US", "2015550123"))
            assertEquals("+1 201-555-0123", formatInternational("US", "2015550123"))
            assertEquals("tel:+1-201-555-0123", formatRfc3966("US", "2015550123"))
        }
        check("PhoneNumber typealias reads fields") {
            val n: PhoneNumber = parse("+1 201 555 0123", "US")
            assertTrue("valid parse", n.error().isEmpty())
        }

        // ---- report ----
        println()
        if (failures.isEmpty()) {
            println("PASS (v2) — $passed checks")
            System.exit(0)
        }
        println("FAIL (v2) — ${failures.size} of ${passed + failures.size} checks failed:")
        failures.forEach { println("  $it") }
        System.exit(1)
    }

    private fun check(name: String, body: () -> Unit) {
        try {
            body()
            passed++
            println("  ok   $name")
        } catch (t: Throwable) {
            failures += "$name: $t"
            println("  FAIL $name: $t")
        }
    }

    private fun assertEquals(expected: Any?, actual: Any?) {
        if (expected.toString() != actual.toString()) {
            throw AssertionError("expected <$expected> but was <$actual>")
        }
    }

    private fun assertTrue(what: String, cond: Boolean) {
        if (!cond) throw AssertionError("expected $what")
    }
}

File: clojure/test/org/libphonenumber/ae/conformance_test.clj
(ns org.libphonenumber.ae.conformance-test
  "The 40-check binding conformance suite (docs/conformance.md, v3), in Clojure.

  Proves the **Clojure layer** reaches the same engine behaviour the Java and
  Python suites see. Since that layer sits on the Java binding rather than on
  its own FFI, what this really pins down is that the Clojure wrapping — the
  keyword <-> enum translation, the vector coercion of `regions`/`find-numbers`,
  the plain-map `parse` — marshals every value shape correctly.

  It is NOT a phone-number test suite — the behavioural cases live in the
  engine's own tests and run once, in Aether.

  `clojure.test` rather than an external framework, so the run needs nothing but
  the Clojure jar itself: it works offline and cannot fail resolving a
  test-framework artifact."
  (:require [clojure.test :refer [deftest is run-tests]]
            [org.libphonenumber.ae.core :as pn]))

;; ---- the 40 required checks ----------------------------------------------

(deftest test-01-country-code-us
  (is (= "1" (pn/country-code "US"))))

(deftest test-02-country-code-gb
  (is (= "44" (pn/country-code "GB"))))

(deftest test-03-unknown-region
  (is (= "" (pn/country-code "ZZ"))))

(deftest test-04-example-number
  (is (= "2015550123" (pn/example-number "US"))))

(deftest test-05-possible-lengths
  (is (= "10" (pn/possible-lengths "US"))))

(deftest test-06-region-code-for-country-code
  (is (= "GB" (pn/region-code-for-country-code "44"))))

(deftest test-07-nanpa-country
  (is (true? (pn/nanpa-country? "US"))))

(deftest test-08-region-enumeration
  (let [regs (pn/regions)]
    (is (>= (count regs) 200))
    (is (= 2 (count (first regs))))))

(deftest test-09-regions-for-country-code
  (is (= "US" (first (pn/regions-for-country-code "1")))))

(deftest test-10-14-parse
  (let [p (pn/parse "+1 201 555 0123 ext 42" "US")]
    (is (= "" (:error p)))
    (is (= "2015550123" (:national-number p)))
    (is (= "42" (:extension p)))
    (is (= "1" (:country-code p)))
    (is (= :from-number-with-plus (:source p)))
    (is (= "US" (:region-code p)))))

(deftest test-15-parse-trunk-prefix
  (is (= "1212345678" (:national-number (pn/parse "01212345678" "GB")))))

(deftest test-16-is-possible
  (is (true? (pn/possible-number? "US" "2015550123"))))

(deftest test-17-reason-too-short
  (is (= :too-short (pn/possible-number-reason "US" "201555"))))

(deftest test-18-is-valid
  (is (true? (pn/valid-number? "US" "2015550123"))))

(deftest test-19-invalid-shape
  (is (false? (pn/valid-number? "US" "1015550123"))))

(deftest test-20-valid-with-cc
  (is (true? (pn/valid-number? "US" "+12015550123"))))

(deftest test-21-number-type
  ;; US fixedLine==mobile -> :fixed-line-or-mobile; GB has distinct patterns.
  (is (= :fixed-line-or-mobile (pn/number-type "US" "2015550123")))
  (is (= :fixed-line (pn/number-type "GB" "2070313000"))))

(deftest test-22-format-national
  (is (= "(201) 555-0123" (pn/format-number "US" "2015550123" :national))))

(deftest test-23-format-e164
  (is (= "+12015550123" (pn/format-number "US" "2015550123" :e164))))

(deftest test-24-format-international
  (is (= "+1 201-555-0123" (pn/format-number "US" "2015550123" :international))))

(deftest test-25-format-rfc3966
  (is (= "tel:+1-201-555-0123" (pn/format-number "US" "2015550123" :rfc3966))))

(deftest test-26-match-exact
  (is (= :exact (pn/number-match "+12015550123" "+1 201 555 0123"))))

(deftest test-27-match-none
  (is (= :no-match (pn/number-match "+12015550123" "+12025550123"))))

(deftest test-28-normalize
  (is (= "12015550123" (pn/normalize-digits-only "+1 (201) 555.0123"))))

(deftest test-29-alpha
  (is (= "1-800-3569377" (pn/convert-alpha-characters "1-800-FLOWERS"))))

(deftest test-30-truncate
  (is (= "2015550123" (pn/truncate-too-long "US" "20155501239999"))))

(deftest test-31-as-you-type
  (is (= "(201) 555-0123" (pn/as-you-type "US" "2015550123"))))

(deftest test-32-find-numbers-count
  (is (= 2 (count (pn/find-numbers "call 201-555-0123 or +1 202 555 0199" "US" :valid)))))

(deftest test-33-find-numbers-raw
  (is (= "201-555-0123" (:raw (first (pn/find-numbers "call 201-555-0123 now" "US" :valid))))))

(deftest test-34-abi-version
  (is (= 3 (pn/abi-version))))

(deftest test-35-emergency-us-911
  (is (true? (pn/emergency-number? "US" "911"))))

(deftest test-36-not-emergency-us-999
  (is (false? (pn/emergency-number? "US" "999"))))

(deftest test-37-emergency-gb-999
  (is (true? (pn/emergency-number? "GB" "999"))))

(deftest test-38-valid-short-us-911
  (is (true? (pn/valid-short-number? "US" "911"))))

(deftest test-39-expected-cost-us-911
  (is (= :toll-free (pn/short-expected-cost "US" "911"))))

(deftest test-40-short-example-us
  (is (= "112" (pn/short-example-number "US"))))

;; ---- extras specific to the Clojure layer --------------------------------

(deftest test-format-default-style-is-national
  (is (= "(201) 555-0123" (pn/format-number "US" "2015550123"))))

(deftest test-format-helpers-agree
  (is (= "+12015550123" (pn/format-e164 "US" "2015550123")))
  (is (= "+1 201-555-0123" (pn/format-international "US" "2015550123")))
  (is (= "tel:+1-201-555-0123" (pn/format-rfc3966 "US" "2015550123")))
  (is (= "(201) 555-0123" (pn/format-national "US" "2015550123"))))

;; ---- runner --------------------------------------------------------------

(defn -main
  "Entry point for `clojure.main -m`, so the suite runs with nothing but the
  Clojure jar. Exits non-zero if anything failed, which is what the aeb node
  keys off."
  [& _args]
  (let [{:keys [fail error]} (run-tests 'org.libphonenumber.ae.conformance-test)]
    (shutdown-agents)
    (System/exit (if (pos? (+ fail error)) 1 0))))
